language/diem-vm/mvhashmap/src/unit_tests/proptest_types.rs

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::MVHashMap;
+use proptest::{collection::vec, prelude::*, sample::Index, strategy::Strategy};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::Debug,
+    hash::Hash,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+const DEFAULT_TIMEOUT: u64 = 30;
+
+#[derive(Debug, Clone)]
+enum Operator<V: Debug + Clone> {
+    Insert(V),
+    Remove,
+    Skip,
+    Read,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum ExpectedOutput<V: Debug + Clone + PartialEq> {
+    NotInMap,
+    Deleted,
+    Value(V),
+}
+
+struct Baseline<K, V>(HashMap<K, BTreeMap<usize, Option<V>>>);
+
+impl<K, V> Baseline<K, V>
+where
+    K: Hash + Eq + Clone,
+    V: Clone + Debug + PartialEq,
+{
+    pub fn new(txns: &[(K, Operator<V>)]) -> Self {
+        let mut baseline: HashMap<K, BTreeMap<usize, Option<V>>> = HashMap::new();
+        for (idx, (k, op)) in txns.iter().enumerate() {
+            let value_to_update = match op {
+                Operator::Insert(v) => Some(v.clone()),
+                Operator::Remove => None,
+                Operator::Skip | Operator::Read => continue,
+            };
+
+            baseline
+                .entry(k.clone())
+                .or_insert_with(BTreeMap::new)
+                .insert(idx, value_to_update);
+        }
+        Self(baseline)
+    }
+
+    pub fn get(&self, key: &K, version: usize) -> ExpectedOutput<V> {
+        match self
+            .0
+            .get(key)
+            .and_then(|tree| tree.range(..version).last())
+        {
+            None => ExpectedOutput::NotInMap,
+            Some((_, Some(v))) => ExpectedOutput::Value(v.clone()),
+            Some((_, None)) => ExpectedOutput::Deleted,
+        }
+    }
+}
+
+fn operator_strategy<V: Arbitrary + Clone>() -> impl Strategy<Value = Operator<V>> {
+    prop_oneof![
+        2 => any::<V>().prop_map(Operator::Insert),
+        1 => Just(Operator::Remove),
+        1 => Just(Operator::Skip),
+        4 => Just(Operator::Read),
+    ]
+}
+
+fn run_and_assert<K, V>(
+    universe: Vec<K>,
+    transaction_gens: Vec<(Index, Operator<V>)>,
+) -> Result<(), TestCaseError>
+where
+    K: PartialOrd + Send + Clone + Hash + Eq + Sync,
+    V: Send + Debug + Clone + PartialEq + Sync,
+{
+    let transactions: Vec<(K, Operator<V>)> = transaction_gens
+        .into_iter()
+        .map(|(idx, op)| (idx.get(&universe).clone(), op))
+        .collect::<Vec<_>>();
+
+    let versions_to_write = transactions
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, (key, op))| match op {
+            Operator::Read => None,
+            Operator::Insert(_) | Operator::Skip | Operator::Remove => Some((key.clone(), idx)),
+        })
+        .collect::<Vec<_>>();
+
+    let baseline = Baseline::new(transactions.as_slice());
+    let (_, map) = MVHashMap::<K, Option<V>>::new_from_parallel(versions_to_write);
+    let curent_idx = AtomicUsize::new(0);
+
+    rayon::scope(|s| {
```

Comment:
can you add a few comments to the test to explain the logic? just a few lines at key places should suffice.